src/net/server/handlers/query.rs

```diff
@@ -64,13 +65,34 @@ impl<T> Clone for ReservedPermit<T> {
 
 /// Middleware that first reserves a permit on the channel to send messages to the messaging layer.
 /// Once reserved, adds the permit to the extension for retrieval from the handler.
-pub async fn obtain_permit_mw<T: Send + 'static, B>(
-    sender: mpsc::Sender<T>,
-    mut req: Request<B>,
+/// # Panics
+/// if messages arrive out of order
+pub async fn obtain_permit_mw<B: Send>(
+    req: Request<B>,
     next: Next<B>,
 ) -> Result<Response, MpcHelperServerError> {
+    let mut req_parts = RequestParts::new(req);
+    let Path(query_id, step) = Path::from_request(&mut req_parts).await?;
+    let Query(RoleQueryParam { role }) =
+        Query::<RoleQueryParam>::from_request(&mut req_parts).await?;
+    let record_headers = RecordHeaders::from_request(&mut req_parts).await?;
+    let Extension(last_seen_messages) =
+        Extension::<LastSeenMessages>::from_request(&mut req_parts).await?;
+    // PANIC if messages arrive out of order; pretty print the error
+    // TODO (ts): remove this when streaming solution is complete
+    last_seen_messages.update_in_place(&ChannelId::new(role, step), record_headers.offset);
+
+    let Extension(message_send_map) =
+        Extension::<MessageSendMap>::from_request(&mut req_parts).await?;
+
+    let sender = message_send_map.get(query_id)?;
```

Comment:
Storing `HttpNetwork` inside the map is a big step towards the proper encapsulation. I think what we would want here (probably in a later PR) is hide the sender channel from the axum mw/handler layer as well. It could be as simple as exposing a wrapper `Permit` struct and a method `reserve` on `HttpNetwork` that returns it. In this case middleware does not need to know how `HttpNetwork` is implemented